### ISSUE TITLE
WebUI shows error when trying to open AAS and running behind nginx HTTPS Reverse Proxy

### ISSUE THREAD

**Issue by pjaydev** [eclipse-basyx/basyx-applications#109]
Hello,

we are currently testing out the AAS GUI / AAS WebUI for potential internal use cases.

For this purpose, I installed and modified the examples docker-compose.yml stack taken from https://github.com/eclipse-basyx/basyx-java-server-sdk/tree/main/examples via docker compose on our server, which is publicly accessible (I replaced our original domain by aas.my-domain.com for demonstration in this issue).

We use a separately running **nginx** instance as HTTPS Gateway to publicly access the services via Reverse Proxy.

When running the docker compose services stack locally on my dev system without HTTPS running at http://localhost:3000 as suggested, everything works fine.

But after deploying and integrating the stack on our server, when trying to open any of the listed AAS's, I see the following error message:

![errormessage](https://github.com/eclipse-basyx/basyx-applications/assets/34743592/91690ed8-f6da-47ad-8bdd-491aaed35ab4)

These are the modifications I made in the `docker-compose.yml`:

```
...
services:
	...
	aas-web-ui_v2:
	...
	ports:
      - "9000:3000"
	environment:
      AAS_REGISTRY_PATH: "https://aas.my-domain.com/aas-registry"
      SUBMODEL_REGISTRY_PATH: "https://aas.my-domain.com/sm-registry"
      AAS_DISCOVERY_PATH: "https://aas.my-domain.com/aas-discovery"
      AAS_REPO_PATH: "https://aas.my-domain.com/aas-env/shells"
      SUBMODEL_REPO_PATH: "https://aas.my-domain.com/aas-env/submodels"
      CD_REPO_PATH: "https://aas.my-domain.com/aas-env/concept-descriptions"
...
```

Additionally, for all of the services, iÍ mapped the container ports 808x to the host ports 908x, since we need the 80xx ports for other services).

As you see here, the endpoint configuration was properly applied in the web app:

![endpoint_config](https://github.com/eclipse-basyx/basyx-applications/assets/34743592/08fc6e29-0709-4b19-9711-b91382d5c784)

This is the modification I made in `aas-env.properties`:

```
...
basyx.externalurl = https://aas.my-domain.com
```

This is my nginx config file for the Reverse Proxy configuration:
```
server {

    client_max_body_size 100M;
    server_name aas.my-domain.com;

    location / {
        proxy_pass http://localhost:9000;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        add_header 'Access-Control-Allow-Origin' '*';
    }

    location /aas-env/ {
		# This rewrite rule converts requests like https://aas.my-domain.com/aas-env/foo/bar to 
		# https://localhost:9081/foo/bar to pretend root path requests for the API endpoints:
        rewrite ^/aas-env/(.*)$ /$1 break; 

        proxy_pass http://localhost:9081;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        add_header 'Access-Control-Allow-Origin' '*';
    }

    location /aas-registry/ {
        rewrite ^/aas-registry/(.*)$ /$1 break;
        
        proxy_pass http://localhost:9082;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        add_header 'Access-Control-Allow-Origin' '*';
    }

    location /sm-registry/ {
        rewrite ^/sm-registry/(.*)$ /$1 break;
        
        proxy_pass http://localhost:9083;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        add_header 'Access-Control-Allow-Origin' '*';
    }

    location /aas-discovery/ {
        rewrite ^/aas-discovery/(.*)$ /$1 break;
        
        proxy_pass http://localhost:9084;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        add_header 'Access-Control-Allow-Origin' '*';
    }

    access_log /var/log/nginx/aas.my-domain.com.access.log;
    error_log /var/log/nginx/aas.my-domain.com.error.log;

    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/aas.my-domain.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/aas.my-domain.com/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}

server {
    if ($host = aas.my-domain.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


    server_name aas.my-domain.com;

    listen 80;
    return 404; # managed by Certbot

}
```

Unfortunately I can not find out any further details about the error. I checked the log outputs in the browser dev console as well as the logs of all involved docker services, but nowhere any error log output is generated.

Since it's working fine using just HTTP, my assumption that there might be any problems somewhere in the data flow where a HTTPS response could be interpreted as a HTTP response and therefore leading to this type of error, but I also might be wrong here.

Do you have any ideas?  
Thanks in advance!

**Comment by aaronzi** [comment]
Hello @pjaydev,

first of all sorry for the late reply.

I was able to reproduce the issue you have encountered with my own small Nginx Docker example. The problem you encountered was caused by a bug in the registry integration feature of the AAS Environment (AAS and Submodel Repository). This bug was fixed in this [PR](https://github.com/eclipse-basyx/basyx-java-server-sdk/pull/265). The changes are now part of the current versions of the [BaSyx off-the-shelf components on Docker Hub](https://hub.docker.com/layers/eclipsebasyx/aas-environment/2.0.0-SNAPSHOT/images/sha256-9ef7ae78878a205e580d19a495469df16de81b321e7911bfb263a931599069aa?context=explore). Please make sure to delete the old images locally because the tag names did not change!

One thing to look out for right now is that the `basyx.externalurl` should have a trailing slash!

It could look like this:
```
basyx.externalurl = https://my-url/aas-env/
```

Please try the new version and check if it fixes the problem you encountered.

**Comment by pjaydev** [comment]
I also just tried to use `aaronzi/basyx-aas-ui:SNAPSHOT_15` as source for the docker image of the Web UI instead of `eclipsebasyx/aas-gui:v2-240327`. (I discovered this alternative image source in the docker-compose files generated by the BaSyx Starter Kit.)

Using this version of Web UI I get a little bit more detailed error output:

![image](https://github.com/eclipse-basyx/basyx-applications/assets/34743592/59c47eb4-5142-4a74-8154-3377b5a4deef)

and 

![image](https://github.com/eclipse-basyx/basyx-applications/assets/34743592/971b076c-3568-414a-b4fc-ccde0bd683fa)

But I guess this just describes the behaviour caused by the problem you already discovered (wrong endpoint, resulting in an empty response).

**Comment by aaronzi** [comment]
Hi @pjaydev,

I can confirm that the PR is included in the newest Version of the Snapshot image. Could you please check if you configured the registry integration the right way. Your BaSyx AAS Env config should look similar to this:

```
server.port=8081

basyx.backend = InMemory

basyx.environment=file:aas
basyx.cors.allowed-origins=*
basyx.cors.allowed-methods=GET,POST,PATCH,DELETE,PUT,OPTIONS,HEAD

basyx.aasrepository.feature.registryintegration = http://aas-registry:8080/aas-registry
basyx.submodelrepository.feature.registryintegration = http://sm-registry:8080/sm-registry
# basyx.externalurl with trailing slash
basyx.externalurl = http://localhost/aas-env/
server.servlet.context-path = /aas-env
```

I created a new example on how to use BaSyx with nginx. You can find the PR adding this example here: https://github.com/eclipse-basyx/basyx-java-server-sdk/pull/289/files
This maybe provides you with some insights.

Also thank you for reminding me to exchange the image of the AAS UI in the Starter Kit to the official BaSyx image.

**Comment by pjaydev** [comment]
I tried again after comparing mine with your configuration and using the `eclipsebasyx/aas-gui:v2-240515` image of the Web UI. 

I can't really tell what I did differently compared to my last attempt, but this time indeed it seems to work fine!

However I had to **leave out** the line: `server.servlet.context-path = /aas-env` 
When keeping this line, the **aas-env** container never reaches a healthy state.

Thanks again for your help!
